hal/src/main/native/cpp/Main.cpp

```diff
@@ -0,0 +1,48 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "hal/Main.h"
+
+#include <wpi/condition_variable.h>
+#include <wpi/mutex.h>
+
+static void DefaultMain();
+static void DefaultExit();
+
+static bool gHasMain = false;
+static void (*gMainFunc)() = DefaultMain;
+static void (*gExitFunc)() = DefaultExit;
+static bool gExited = false;
+static wpi::mutex gExitMutex;
+static wpi::condition_variable gExitCv;
+
+static void DefaultMain() {
+  std::unique_lock lock{gExitMutex};
+  gExitCv.wait(lock, [] { return gExited; });
+}
+
+static void DefaultExit() {
+  std::lock_guard lock{gExitMutex};
+  gExited = true;
+  gExitCv.notify_all();
+}
+
```

Comment:
We should be using an initializer function to initialize these things, rather then static initialization. Just in case these functions get called in a static initialization context (they shouldn't, but I don't know).

Reply:
It's pretty unlikely, but I moved the mutex and cv into an initializer anyway